driver/src/main/java/org/neo4j/driver/internal/metrics/MicrometerConnectionPoolMetrics.java

```diff
@@ -0,0 +1,277 @@
+    private final Counter acquired;
+    private final Counter timedOutToAcquire;
+    private final Timer totalAcquisitionTimer;
+    private final Timer totalConnectionTimer;
+    private final Timer totalInUseTimer;
+    private final Counter released;
+
+    MicrometerConnectionPoolMetrics( String poolId, ServerAddress address, IntSupplier inUseSupplier, IntSupplier idleSupplier, MeterRegistry registry )
+    {
+        this( poolId, address, inUseSupplier, idleSupplier, registry, Tags.empty() );
+    }
+
+    MicrometerConnectionPoolMetrics( String poolId, ServerAddress address, IntSupplier inUseSupplier, IntSupplier idleSupplier, MeterRegistry registry, Iterable<Tag> initialTags )
+    {
+        Objects.requireNonNull( poolId );
+        Objects.requireNonNull( address );
+        Objects.requireNonNull( inUseSupplier );
+        Objects.requireNonNull( idleSupplier );
+        Objects.requireNonNull( registry );
+
+        this.id = poolId;
+        this.inUseSupplier = inUseSupplier;
+        this.idleSupplier = idleSupplier;
+        String host = address instanceof BoltServerAddress ? ((BoltServerAddress) address).connectionHost() : address.host();
+        Iterable<Tag> tags = Tags.concat( initialTags,
+                                          "address", String.format( "%s:%d", host, address.port() ) );
+
+        Gauge.builder( IN_USE, this::inUse ).tags( tags ).register( registry );
+        Gauge.builder( IDLE, this::idle ).tags( tags ).register( registry );
+        Gauge.builder( CREATING, creating, AtomicInteger::get ).tags( tags ).register( registry );
+        created = Counter.builder( CREATED ).tags( tags ).register( registry );
+        failedToCreate = Counter.builder( FAILED ).tags( tags ).register( registry );
+        closed = Counter.builder( CLOSED ).tags( tags ).register( registry );
+        Gauge.builder( ACQUIRING, acquiring, AtomicInteger::get ).tags( tags ).register( registry );
+        acquired = Counter.builder( ACQUIRED ).tags( tags ).register( registry );
+        timedOutToAcquire = Counter.builder( ACQUISITION_TIMEOUT ).tags( tags ).register( registry );
+        totalAcquisitionTimer = Timer.builder( ACQUISITION ).tags( tags ).register( registry );
+        totalConnectionTimer = Timer.builder( CREATION ).tags( tags ).register( registry );
+        totalInUseTimer = Timer.builder( USAGE ).tags( tags ).register( registry );
+        released = Counter.builder( RELEASED ).tags( tags ).register( registry );
+    }
+
+    @Override
+    public void beforeCreating( ListenerEvent<?> connEvent )
+    {
+        creating.incrementAndGet();
+        connEvent.start();
+    }
+
+    @Override
+    public void afterFailedToCreate()
+    {
+        failedToCreate.increment();
+        creating.decrementAndGet();
+    }
+
+    @Override
+    public void afterCreated( ListenerEvent<?> connEvent )
+    {
+        creating.decrementAndGet();
+        created.increment();
```

Comment:
Timers include a count, so a separate Counter is not needed if it counts the same events being timed.

Reply:
Good point, thanks.